src/axolotl/utils/config/models/input/v0_4_1/__init__.py

```diff
@@ -0,0 +1,931 @@
+
+    special_tokens: Optional[SpecialTokensConfig] = None
+    tokens: Optional[List[str]] = None
+
+    torch_compile: Optional[bool] = None
+    torch_compile_backend: Optional[str] = None
+
+    max_steps: Optional[int] = None
+    warmup_steps: Optional[int] = None
+    warmup_ratio: Optional[float] = None
+    eval_steps: Optional[int] = None
+    evaluation_strategy: Optional[str] = None
+    save_steps: Optional[int] = None
+    saves_per_epoch: Optional[int] = None
+    save_strategy: Optional[str] = None
+    save_total_limit: Optional[int] = None
+    logging_steps: Optional[int] = None
+    early_stopping_patience: Optional[int] = None
+
+    neftune_noise_alpha: Optional[float] = None
+
+    max_memory: Optional[Union[int, str]] = None
+    gpu_memory_limit: Optional[str] = None
+
+    chat_template: Optional[Union[str, ChatTemplate]] = None
+    default_system_message: Optional[str] = None
+
+    # INTERNALS - document for now
+    total_num_tokens: Optional[int] = None
+    total_supervised_tokens: Optional[int] = None
+    sample_packing_eff_est: Optional[float] = None
+    axolotl_config_path: Optional[str] = None
+
+    @field_validator("datasets")
+    @classmethod
+    def check_non_empty_datasets(cls, datasets):
+        if len(datasets) == 0:
+            raise ValueError("datasets list cannot be empty")
+        return datasets
+
+    @field_validator("datasets", mode="before")
+    @classmethod
+    def fix_sharegpt_datasets(cls, datasets):
+        for idx, ds_cfg in enumerate(datasets):
+            if not ds_cfg["type"]:
+                continue
+            if ds_cfg["type"] == "sharegpt:chat":
+                LOG.warning(
+                    PendingDeprecationWarning(
+                        "`type: sharegpt:chat` will soon be deprecated. simply use `type: sharegpt` instead."
+                    )
+                )
+                datasets[idx]["type"] = "sharegpt"
+            if "sharegpt_simple" in ds_cfg["type"]:
+                LOG.warning(
+                    PendingDeprecationWarning(
+                        "`type: sharegpt_simple` will soon be deprecated. simply use `type: sharegpt` instead."
+                    )
+                )
+                datasets[idx]["type"] = datasets[idx]["type"].replace(
+                    "sharegpt_simple", "sharegpt"
+                )
+        return datasets
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_batch_size_fields(cls, root):
+        non_empty_count = sum(
+            1
+            for field in (
+                "micro_batch_size",
+                "gradient_accumulation_steps",
+                "batch_size",
+            )
+            if root.get(field)
+        )
+        if non_empty_count < 2:
+            raise ValueError(
+                "At least two of [micro_batch_size, gradient_accumulation_steps, batch_size] must be set"
+            )
+        return root
```

Comment:
Just a slight refactor.
```suggestion
    fields = ("micro_batch_size", "gradient_accumulation_steps", "batch_size")
    non_empty_count = sum(1 for field in fields if root.get(field))
    
    if non_empty_count < 2:
        raise ValueError("At least two of {} must be set".format(', '.join(fields)))
    
    return root
```